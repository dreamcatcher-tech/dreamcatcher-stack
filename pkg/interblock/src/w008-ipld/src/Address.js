import assert from 'assert-fast'
import { encode, prepare } from '@ipld/dag-pb'
import { CID } from 'multiformats/cid'
import { Block } from 'multiformats/block'
import { IpldInterface } from './IpldInterface'
import { sha256 } from 'multiformats/hashes/sha2'
import { Pulse } from '.'

export const cidV0FromString = (string) => {
  assert.strictEqual(typeof string, 'string')
  const bytes = encode(prepare(string))
  const hash = sha256.digest(bytes)
  const cid = CID.createV0(hash)
  return cid
}
export const generateConstantCids = () => {
  const UNKNOWN = cidV0FromString('UNKNOWN')
  const ROOT = cidV0FromString('ROOT')
  const LOOPBACK = cidV0FromString('LOOPBACK')
  const INVALID = cidV0FromString('INVALID')
  const GENESIS = cidV0FromString('GENESIS')
  const IO = cidV0FromString('IO')
  console.log({ UNKNOWN, ROOT, LOOPBACK, INVALID, GENESIS, IO })
}
// generateConstantCids()

const UNKNOWN = CID.parse('QmVQEFQi81SSbZaezrm78p333EGeYzfEZyvCnX848KaMCw')
const ROOT = CID.parse('QmSawJHmTNpaUjDYWCE9RgoHTpKbi6JD7TuGESFbtZ4ZLc')
const LOOPBACK = CID.parse('Qme2gGBx8EnSrXc5shQF867KPQd4jwubNov67KEKZbo4p3')
const INVALID = CID.parse('QmYSWwmJ4w1pZ6igGRNKcVHpBU68iaumYEjsdbMpxfAQaj')
const GENESIS = CID.parse('QmZTKF2kuFHy8isKWXpNeNa5zjeJwsHUbPbTNF1fS8HkpB')
const IO = CID.parse('QmapFBxqMFEFxSqGE45yVYKid471NoegxSriH1GpCMfUa6')
const defines = [UNKNOWN, ROOT, LOOPBACK, INVALID, GENESIS, IO]

const addressBlock = (cidV1) => {
  assert(cidV1 instanceof CID)
  assert.strictEqual(cidV1.version, 1)
  const value = prepare({ Links: [cidV1] })
  const bytes = encode(value)
  const hash = sha256.digest(bytes)
  const cid = CID.createV0(hash)
  assert.strictEqual(cid.version, 0)
  return new Block({ cid, bytes, value })
}

export class Address extends IpldInterface {
  #cid
  #ipldBlock
  static createCI(string) {
    assert.strictEqual(typeof string, 'string')
    assert(string)
    const value = prepare({ Data: string })
    const bytes = encode(value)
    const hash = sha256.digest(bytes)
    const cid = CID.createV0(hash)
    assert.strictEqual(cid.version, 0)
    const block = new Block({ cid, bytes, value })
    return Address.#createFromBlock(block)
  }
  static createUnknown() {
    return this.#createPredefined(UNKNOWN)
  }
  static createRoot() {
    return this.#createPredefined(ROOT)
  }
  static createLoopback() {
    return this.#createPredefined(LOOPBACK)
  }
  static createInvalid() {
    return this.#createPredefined(INVALID)
  }
  static createGenesis() {
    return this.#createPredefined(GENESIS)
  }
  static createIo() {
    return this.#createPredefined(IO)
  }
  static generate(pulse) {
    assert(pulse instanceof Pulse)
    assert(!pulse.isModified(), `Pulse must be crushed already`)
    assert(pulse.isGenesis(), `Pulse must be genesis`)
    // TODO check the pulse is genesis
    const block = addressBlock(pulse.cid)
    return this.#createFromBlock(block)
  }
  static #createFromBlock(block) {
    assert(block instanceof Block)
    const instance = new this()
    instance.#setCid(block.cid)
    instance.#ipldBlock = block
    return instance
  }
  static #createPredefined(cid) {
    assert(defines.includes(cid), `not predefined`)
    const instance = new this()
    instance.#setCid(cid)
    return instance
  }
  #setCid(cid) {
    assert(cid instanceof CID, `cid must be a CID, got ${cid}`)
    assert(cid.version === 0)
    this.#cid = cid
  }
  static async uncrush(rootCid, resolver, options) {
    assert(rootCid instanceof CID)
    assert.strictEqual(typeof resolver, 'function')
    for (const define of defines) {
      if (define.equals(rootCid)) {
        return this.#createPredefined(define)
      }
    }
    const block = await resolver(rootCid)
    assert(block instanceof Block)
    const instance = this.#createFromBlock(block)
    instance.#ipldBlock = undefined
    return instance
  }
  isModified() {
    return false
  }
  get ipldBlock() {
    return this.#ipldBlock
  }
  get cid() {
    return this.#cid
  }
  [Symbol.for('nodejs.util.inspect.custom')]() {
    return this.toString()
  }
  toString() {
    return `Address(${this.cid.toString().substring(0, 14)})`
  }
  crush() {
    const instance = new Address()
    instance.#setCid(this.#cid)
    return instance
  }
  getDiffBlocks() {
    const map = new Map()
    if (this.ipldBlock) {
      map.set(this.cid, this.ipldBlock)
    }
    return map
  }
  equals(address) {
    assert(address instanceof Address)
    return this.cid.equals(address.cid)
  }
  getChainId() {
    if (this.#cid === LOOPBACK) {
      return 'LOOPBACK'
    }
    if (this.#cid === ROOT) {
      return 'ROOT'
    }
    if (!this.isResolved()) {
      throw new Error(`Address not resolved: ${this.#cid}`)
    }
    return this.#cid.toString()
  }
  // TODO broaden isResolved to cover root and loopback cases
  isRoot() {
    return this.#cid === ROOT
  }
  isLoopback() {
    return this.#cid === LOOPBACK
  }
  isGenesis() {
    return this.#cid === GENESIS
  }
  isUnknown() {
    return this.#cid === UNKNOWN
  }
  isInvalid() {
    return this.#cid === INVALID
  }
  isResolved() {
    return !this.isUnknown() && !this.isInvalid() && !this.isGenesis()
  }
  isRemote() {
    return this.isResolved() && !this.isRoot() && !this.isLoopback()
  }
}
