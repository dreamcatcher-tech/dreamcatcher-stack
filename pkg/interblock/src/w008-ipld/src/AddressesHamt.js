import { Hamt } from './Hamt'
import { Address } from '.'
import assert from 'assert-fast'
export class AddressesHamt extends Hamt {
  get isBakeSkippable() {
    return true
  }
  async set(address, channelId) {
    assert(address instanceof Address)
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(address.isRemote())
    const key = address.getChainId()
    return await super.set(key, channelId)
  }
  async has(address) {
    assert(address instanceof Address)
    assert(address.isRemote())
    return await super.has(address.getChainId())
  }
  async delete(address) {
    assert(await this.has(address), `address not present ${address}`)
    return await super.delete(address.getChainId())
  }
  async get(address) {
    assert(address instanceof Address)
    return await super.get(address.getChainId())
  }
  async compare(other) {
    let { added, deleted, modified } = await super.compare(other)
    added = new Set([...added].map((key) => Address.fromChainId(key)))
    deleted = new Set([...deleted].map((key) => Address.fromChainId(key)))
    modified = new Set([...modified].map((key) => Address.fromChainId(key)))
    return { added, deleted, modified }
  }
}
