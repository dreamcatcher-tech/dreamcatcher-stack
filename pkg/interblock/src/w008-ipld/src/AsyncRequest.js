import assert from 'assert-fast'
import { Request, RequestId, Reply } from '.'
import { IpldStruct } from './IpldStruct'
import equals from 'fast-deep-equal'

/**
    type AsyncRequest struct {
        request &Request
        to String
        id RequestId
        settled optional &Reply
    }
*/
export class AsyncRequest extends IpldStruct {
  static create(request, to) {
    assert(request instanceof Request)
    assert.strictEqual(typeof to, 'string')
    assert(to)
    const instance = new AsyncRequest()
    instance.request = request
    instance.to = to
    return instance
  }
  setId(id) {
    assert(id instanceof RequestId)
    return this.setMap({ id })
  }
  settle(reply) {
    assert(reply instanceof Reply)
    assert(!this.isSettled())
    return this.setMap({ settled: reply })
  }
  isRequestMatch(request) {
    assert(request instanceof Request)
    return equals(this.request, request)
  }
  isSettled() {
    return this.settled !== undefined
  }
  assertLogic() {
    if (this.settled) {
      assert(this.id !== undefined, `cannot settle unidentified requests`)
    }
  }
}
