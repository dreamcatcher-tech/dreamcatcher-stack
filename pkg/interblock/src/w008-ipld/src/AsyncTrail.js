import assert from 'assert-fast'
import { IpldStruct } from './IpldStruct'
import {
  Pulse,
  RxReply,
  AsyncRequest,
  Reply,
  Request,
  RxRequest,
  RequestId,
} from '.'
/**
 * Lifecycle increases thru five states:
 * 1. origin      - before any reduction has taken place
 * 2. reduced     - having txs with no requestIds attached
 * 3. transmitted - all txs have requestIds attached
 * 4. fulfilled   - all txs have been settled with replies
 * 5. settled     - the trail has replied to the origin
 * 
 * 
type RequestId struct {
    channelId Int
    stream String
    requestIndex Int
}
type AsyncRequest struct {
    request &Request
    to String
    id RequestId
    reply optional &Reply
}
type RxRequest struct {
    request &Request
    requestId RequestId
}
type AsyncTrail struct {
    origin RxRequest
    settles [AsyncRequest]
    txs [AsyncRequest]
    rxReply optional &RxReply 
}
 */
export class AsyncTrail extends IpldStruct {
  settles = []
  txs = []
  static createCI() {
    const requestId = RequestId.createCI()
    const request = Request.create('TEST')
    const rxRequest = RxRequest.create(request, requestId)
    return this.create(rxRequest)
  }
  static create(origin) {
    assert(origin instanceof RxRequest)
    const instance = new this()
    Object.assign(instance, { origin })
    return instance
  }
  static createWithPulse(origin, pulse) {
    /** used when we need the response pattern, but want to make
     * one time only synchronized changes to a state variable.
     * If this mode is used, settles are disallowed, and txs
     * are disallowed, as the prescence of these items means
     * repeated execution will occur, which is counter to purpose
     * of this function.
     */
    assert(origin instanceof RxRequest)
    assert(pulse instanceof Pulse)
    return super.clone({ origin, pulse })
  }
  isPending() {
    return !this.rxReply
  }
  assertLogic() {
    assert(!this.pulse)
    assert(!this.txs.every((tx) => tx.isSettled()))
    assert(this.settles.every((tx) => tx.isSettled()))
    assert(!this.rxReply, `can only crush promised trails`)
  }
  isFulfilled() {
    // check that all settles are in fact settled
    return !this.txs.length && this.settles.every((s) => s.isSettled())
  }
  setTxs(txs) {
    assert(Array.isArray(txs))
    assert(txs.every((tx) => tx instanceof AsyncRequest))
    assert(txs.every((tx) => !tx.isSettled()))
    return this.setMap({ txs })
  }
  updateTxs(txs) {
    // used for adding requestIds to txs
    assert(Array.isArray(txs))
    assert(this.txs.length)
    assert.strictEqual(this.txs.length, txs.length)
    // TODO check requests match
    return this.setTxs(txs)
  }
  settleOrigin(reply) {
    assert(reply instanceof Reply)
    assert(!this.rxReply)
    const rxReply = RxReply.create(reply, this.origin.requestId)
    return this.setMap({ rxReply })
  }
  settleTx(rxReply) {
    assert(rxReply instanceof RxReply)
    let isMatched = false
    let txs = this.txs.map((tx) => {
      if (tx.isIdMatch(rxReply)) {
        isMatched = true
        return tx.settle(rxReply)
      }
      return tx
    })
    assert(isMatched, `no match found for ${rxReply.requestId}`)
    let { settles } = this
    if (txs.every((tx) => tx.isSettled())) {
      settles = settles.concat(txs)
      txs = []
    }
    return this.setMap({ txs, settles })
  }
  getSettles() {
    return [...this.settles]
  }
  getRequestObject() {
    return this.origin.getRequestObject()
  }
  setError(txs, error) {
    const reply = Reply.createError(error)
    return this.setTxs(txs).settleOrigin(reply)
  }
  getError() {
    assert(this.rxReply)
    assert(this.rxReply.reply.isRejection())
    return this.rxReply.reply.getRejectionError()
  }
  result() {
    if (this.isPending()) {
      return
    }
    const { reply } = this.rxReply
    if (reply.isResolve()) {
      return reply.payload
    } else {
      throw reply.getRejectionError()
    }
  }
  isTransmitted() {
    assert(!this.pulse)
    assert(this.settles.every((tx) => tx.isSettled()))
    return this.txs.every((tx) => !!tx.requestId)
  }
}
