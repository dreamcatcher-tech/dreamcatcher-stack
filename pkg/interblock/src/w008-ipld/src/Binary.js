import assert from 'assert-fast'
import { CID } from 'multiformats/cid'
import { encode, Block } from 'multiformats/block'
import { IpldInterface } from './IpldInterface'
import * as codec from 'multiformats/codecs/raw'
import { sha256 as hasher } from 'multiformats/hashes/sha2'

export class Binary extends IpldInterface {
  // TODO use the unixFS API so these objects are browseable
  #ipldBlock
  static async create(value) {
    assert(value instanceof Uint8Array)
    // TODO move to factory
    const ipldBlock = await encode({ value, codec, hasher })
    assert.strictEqual(ipldBlock.value, value)
    return this.#create(ipldBlock)
  }
  static #create(ipldBlock) {
    assert(ipldBlock instanceof Block, `#create requires a Block`)
    assert.strictEqual(ipldBlock.value, ipldBlock.bytes)
    const instance = new this()
    instance.#ipldBlock = ipldBlock
    instance.deepFreeze()
    IpldInterface.deepFreeze(ipldBlock)
    return instance
  }
  static async uncrush(rootCid, resolver, options) {
    assert(rootCid instanceof CID, `rootCid must be a CID, got ${rootCid}`)
    assert(typeof resolver === 'function', `resolver must be a function`)
    const ipldBlock = await resolver(rootCid)
    return this.#create(ipldBlock)
  }
  isModified() {
    // TODO verify that the Uint8Array cannot be altered
    return false
  }
  get ipldBlock() {
    return this.#ipldBlock
  }
  crush() {
    return this
  }
  getDiffBlocks(from) {
    if (this === from) {
      return new Map()
    }
    const blocks = new Map()
    blocks.set(this.ipldBlock.cid.toString(), this.ipldBlock)
    return blocks
  }
}
