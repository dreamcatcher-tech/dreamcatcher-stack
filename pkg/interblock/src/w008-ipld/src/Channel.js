import assert from 'assert-fast'
import {
  RxReply,
  RequestId,
  Network,
  Reply,
  Address,
  Rx,
  Tx,
  Request,
  Interpulse,
} from '.'
import Debug from 'debug'
import { IpldStruct } from './IpldStruct'
import { RxRequest } from './RxRequest'
const debug = Debug('interblock:models:channel')

/**
## Channel

`tip` matches up with precedent on the other side.

Tx and Rx are split out so that they can be stored on the block separately.
Tx needs to be separate so that remote fetching does not expose the channelId.
Rx needs to be separate to allow the ingestion of messages to be halted at any point.
Both are separate from Channel so that the potentially large HAMT that stores
all the Channel instances is updated as little as possible, whilst providing rapid
lookup to get channels that are active.

Tx needs to be hashed as it is an independent transmission, but Rx does not
need to be hashed.

The structure implements the design goal of making the Pulse be the context
of the state machine that processes all the actions.

Channel stores rx and tx only after all the activity has been wrung out of them.

```sh
type Channel struct {
    address Address                 # The remote chainId
    tx &Tx
    rx Rx
    aliases [String]                # reverse lookup
}
```
 */

export class Channel extends IpldStruct {
  static cidLinks = ['tx', 'address']
  static classMap = { tx: Tx, rx: Rx }
  static create(channelId, address = Address.createUnknown()) {
    assert(Number.isInteger(channelId))
    assert(channelId >= 0)
    assert(address instanceof Address)
    if (address.isRoot()) {
      assert.strictEqual(channelId, Network.FIXED_IDS.PARENT, `root not parent`)
    }

    const rx = Rx.create()
    const tx = Tx.create()
    return super.clone({ address, rx, tx, aliases: [], channelId })
  }
  static createRoot() {
    const root = Address.createRoot()
    return Channel.create(Network.FIXED_IDS.PARENT, root)
  }
  static createIo() {
    const address = Address.createIo()
    return Channel.create(Network.FIXED_IDS.IO, address)
    // TODO stop Io from being aliased
  }
  static createLoopback() {
    const address = Address.createLoopback()
    return Channel.create(Network.FIXED_IDS.LOOPBACK, address)
  }
  // TODO reject any attempts to alias loopback
  resolve(address) {
    assert(address instanceof Address)
    assert(address.isRemote())
    assert(this.isUnknown(), `Can only resolve unknown channels`)
    return this.setMap({ address })
  }
  isUnknown() {
    return this.address.isUnknown()
  }
  isRemote() {
    return this.address.isRemote()
  }
  assertLogic() {
    const { rx, tx, address, aliases, channelId } = this
    if (this.isUnknown()) {
      assert(rx.isEmpty(), 'Replies to Unknown are impossible')
      assert(!rx.tip)
    }
    if (rx.tip) {
      assert(!address.isLoopback())
    }
    if (address.isLoopback()) {
      assert(!tx.precedent)
      const banned = ['@@OPEN_CHILD']
      const systemRequests = tx.system.requests
      assert(systemRequests.every(({ type }) => !banned.includes(type)))
      assert(!aliases.length)
    }
    if (address.isRoot()) {
      assert.strictEqual(channelId, Network.FIXED_IDS.PARENT, 'Root not parent')
    }
  }
  isNext(channel) {
    assert(channel instanceof Channel)
    // TODO verify the logic follows
    return true
  }
  ingestInterpulse(interpulse) {
    assert(interpulse instanceof Interpulse)
    let { rx } = this
    rx = rx.addTip(interpulse)
    return this.setMap({ rx })
  }
  txRequest(request) {
    assert(request instanceof Request)
    const tx = this.tx.txRequest(request)
    return this.setMap({ tx })
  }
  txSystemReply(reply) {
    assert(reply instanceof Reply)
    const tx = this.tx.txSystemReply(reply)
    const rx = this.rx.shiftSystemRequest()
    return this.setMap({ tx, rx })
  }
  txReducerReply(reply) {
    assert(!this.address.isUnknown(), `Address is unknown`)
    assert(reply instanceof Reply)
    const tx = this.tx.txReducerReply(reply)
    const rx = this.rx.shiftReducerRequest()
    return this.setMap({ tx, rx })
  }
  shiftSystemReply() {
    const rx = this.rx.shiftSystemReply()
    return this.setMap({ rx })
  }
  shiftReducerReply() {
    const rx = this.rx.shiftReducerReply()
    return this.setMap({ rx })
  }
  settlePromise(rxReply) {
    const tx = this.tx.settlePromise(rxReply)
    return this.setMap({ tx })
  }
  rxSystemRequest() {
    const { channelId } = this
    const stream = 'system'
    return this.rx.system.rxRequest(channelId, stream)
  }
  rxSystemReply() {
    const { channelId } = this
    const stream = 'system'
    return this.rx.system.rxReply(channelId, stream)
  }
  rxReducerRequest() {
    const { channelId } = this
    const stream = 'reducer'
    return this.rx.reducer.rxRequest(channelId, stream)
  }
  rxReducerReply() {
    const { channelId } = this
    const stream = 'reducer'
    return this.rx.reducer.rxReply(channelId, stream)
  }
  rxIsEmpty() {
    return this.rx.isEmpty()
  }
  getNextRequestId(request) {
    assert(request instanceof Request)
    const { channelId } = this
    const stream = request.isSystem() ? 'system' : 'reducer'
    const requestIndex = this.tx[stream].requestsLength
    return RequestId.create(channelId, stream, requestIndex)
  }
}
