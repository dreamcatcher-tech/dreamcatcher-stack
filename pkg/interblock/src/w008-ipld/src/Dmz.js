import assert from 'assert-fast'
import {
  Provenance,
  Pulse,
  PulseLink,
  Config,
  Binary,
  Timestamp,
  Network,
  State,
  Meta,
  Pending,
} from '.'
import { IpldStruct } from './IpldStruct'

const ciTimestamp = Timestamp.createCI()
let defaultParams
const getDefaultParams = (CI = false) => {
  if (!defaultParams) {
    defaultParams = {
      config: Config.create(),
      network: Network.create(),
      state: State.create(),
      meta: Meta.create(),
    }
  }
  if (CI) {
    return { ...defaultParams, timestamp: ciTimestamp }
  } else {
    return { ...defaultParams, timestamp: Timestamp.create() }
  }
}
export class Dmz extends IpldStruct {
  static cidLinks = [
    'config',
    'network',
    'state',
    'meta',
    'pending',
    'approot',
    'binary',
  ]
  static classMap = {
    config: Config,
    timestamp: Timestamp,
    network: Network,
    state: State,
    meta: Meta,
    pending: Pending,
    approot: PulseLink,
    binary: Binary,
  }
  static create(params = {}, CI = false) {
    assert.strictEqual(typeof params, 'object')
    params = { ...params }
    const defaultParams = getDefaultParams(CI)
    for (const key in params) {
      assert(this.classMap[key], `key ${key} not mapped to CID class`)
      const isInstanceOf = params[key] instanceof this.classMap[key]
      if (!isInstanceOf) {
        params[key] = defaultParams[key].setMap(params[key])
      }
    }
    params = { ...defaultParams, ...params }
    return super.clone(params)
  }
  assertLogic() {
    // TODO if isSideEffectCapable ensure the validators list is singular
    const { network, config } = this
    // TODO verify that the pending buffers map to legit channels
    // TODO find how to not infect everything with async
    // assert(!network.getIo() || config.isPierced)
  }
}
