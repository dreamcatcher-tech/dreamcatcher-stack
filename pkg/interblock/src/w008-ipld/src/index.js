export { Binary } from './Binary'
export { RequestId } from './RequestId'
export { Request } from './Request'
export { RxRequest } from './RxRequest'
export { Reply } from './Reply'
export { RxReply } from './RxReply'
export { Address } from './Address'
export { PulseLink, HistoricalPulseLink } from './PulseLink'
export { TxQueue } from './TxQueue'
export { Tx } from './Tx'
export { RxQueue } from './RxQueue'
export { Rx } from './Rx'
export { Channel } from './Channel'
export { Config } from './Config'
export { PublicKey } from './PublicKey'
export { Validators } from './Validators'
export { ChannelsHamt } from './ChannelsHamt'
export { ChildrenHamt } from './ChildrenHamt'
export { UplinksHamt } from './UplinksHamt'
export { DownlinksHamt } from './DownlinksHamt'
export { SymlinksHamt } from './SymlinksHamt'
export { AddressesHamt } from './AddressesHamt'
export { Channels } from './Channels'
export { Transmissions } from './Transmissions'
export { Network } from './Network'
export { AsyncRequest } from './AsyncRequest'
export { AsyncTrail } from './AsyncTrail'
export { Pending } from './Pending'
export { State } from './State'
export { Timestamp } from './Timestamp'
export { Dmz } from './Dmz'
export { Provenance } from './Provenance'
export { Interpulse } from './Interpulse'
export { Pulse } from './Pulse'
export { Keypair } from './Keypair'
export { decode } from './IpldUtils'
export { IpldInterface } from './IpldInterface'
export { Hamt } from './Hamt'
