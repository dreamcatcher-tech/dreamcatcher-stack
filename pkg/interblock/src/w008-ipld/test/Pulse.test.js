import { assert } from 'chai/index.mjs'
import {
  Interpulse,
  Request,
  Channel,
  Dmz,
  Provenance,
  Pulse,
  Address,
  Config,
  Keypair,
  Validators,
} from '..'
import { setCiTimestamp } from '../src/Dmz'

let previousLabel
const timer = (label) => {
  if (previousLabel) {
    console.timeEnd(previousLabel)
  }
  if (label) {
    console.time(label)
  }
  previousLabel = label
}

describe('Pulse', () => {
  setCiTimestamp()
  test('basic', async () => {
    let pulse = await Pulse.createCI()
    pulse = await pulse.crush()
    const address = Address.generate(pulse)
    expect(address.cid).toMatchSnapshot()
  })
  test('birth child', async () => {
    timer('create CI')
    let parent = await Pulse.createCI()
    timer('recrush')
    parent = await parent.crush()
    timer()
    const config = { entropy: { seed: 'test' } }
    let { dmz } = parent.provenance
    dmz = await dmz.addChild('child1', { config })
    timer('softpulse')
    parent = await parent.generateGenesisSoftPulse()
    parent = parent.setMap({ provenance: { dmz } })
    timer('crush')
    // when we modify it, it should point the provenance at the previous
    parent = await parent.crush()
    timer()
    assert.throws(() => Address.generate(parent), 'must be genesis')
    timer('keypair')
    const keypair = Keypair.createCI()
    timer('sign')
    let signature = await keypair.sign(parent.provenance)
    timer('add sig')

    assert(!parent.isVerified())
    parent = parent.addSignature(keypair.publicKey, signature)
    assert(!parent.isVerified())
    parent = await parent.crush()
    assert(parent.isVerified())
    timer()

    // generate a genesis pulse using the params in the action
    const { channels } = parent.provenance.dmz.network
    assert.strictEqual(channels.txs.length, 1)
    const [childChannelId] = channels.txs
    const tx = await channels.getTx(childChannelId)
    assert(tx.isGenesisRequest())
    const { validators } = parent.provenance
    const { timestamp } = parent.provenance.dmz
    let child = await tx.extractChildGenesis(validators, timestamp)
    assert(child.isGenesis())
    const { publicKeys } = child.provenance.validators
    assert.strictEqual(publicKeys.length, 1)
    assert.deepEqual(publicKeys[0], keypair.publicKey)
    const channel = await channels.getChannel(childChannelId)
    assert(channel.address.equals(child.getAddress()))

    // reply to the parents action with a signed block
    child = await child.generateGenesisSoftPulse(parent)
    let interpulse = Interpulse.extract(parent, child.getAddress())
    child = await child.ingestInterpulse(interpulse)

    const rxRequest = await child.getNetwork().rxSystemRequest()
    assert.strictEqual(rxRequest.type, '@@GENESIS')
    assert.strictEqual(rxRequest.channelId, 0)
    assert.strictEqual(rxRequest.stream, 'system')
    assert.strictEqual(rxRequest.index, 0)

    // default reply to the request in the child
    let childParentChannel = await child.getNetwork().getParent()
    assert(!childParentChannel.rx.system.isEmpty())
    const network = await child.getNetwork().txSystemReply()
    childParentChannel = await network.getParent()
    assert(childParentChannel.rx.system.isEmpty())
    assert.strictEqual(childParentChannel.tx.system.replies.length, 1)
    assert.strictEqual(network.channels.txs.length, 1)
    assert.strictEqual(network.channels.rxs.length, 0)
    child = child.setNetwork(network)

    // then send it back to the parent
    child = await child.crush()
    signature = await keypair.sign(child.provenance)
    child = child.addSignature(keypair.publicKey, signature)
    child = await child.crush()
    interpulse = Interpulse.extract(child, parent.getAddress())

    parent = await parent.ingestInterpulse(interpulse)
    const rxReply = await parent.getNetwork().rxSystemReply()
    assert.strictEqual(rxReply.type, '@@RESOLVE')
    expect(rxReply).toMatchSnapshot()

    const parentNetwork = await parent.getNetwork().shiftSystemReply()
    parent = parent.setNetwork(parentNetwork)
    parent = await parent.crush()
  })
  test.todo('loopback transmissions')
  test.todo('loopback promises')
  test.todo('receive interpulse')
  test.todo('remove transmission')
  test.todo('softpulse for pooling')
  test.todo('genesis pulse')
  test.todo('pierce pulse')
  test.todo('pending')
  test.todo('public connection')
  test.todo('precedent multiple pulses ahead')
})
