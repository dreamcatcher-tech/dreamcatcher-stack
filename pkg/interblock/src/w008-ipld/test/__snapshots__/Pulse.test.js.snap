// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pulse basic 1`] = `"Address(QmYMKzxgro9NTb)"`;

exports[`Pulse birth child 1`] = `
RxReply {
  "channelId": 3,
  "payload": Object {},
  "requestIndex": 0,
  "stream": "system",
  "type": "@@RESOLVE",
}
`;
